### ISSUE TITLE
Matching, Swaps and Order Invalidation Summary

### ISSUE THREAD

**Issue by moshababo** [ExchangeUnion/xud#277]
Local `placeorder`
* If an internal match was found (maker order is `ownOrder`), send order invalidation for the maker order (#266)
* If a non-internal match was found (maker order is `peerOrder`), swap is instantiated
    * If swap succeed, do nothing
    * If swap failed, try match again

Remote peer `placeorder`
* If a match was found with our order, swap is instantiated by the remote peer
  * If swap succeed, send order invalidation
  * If swap failed, do nothing

**Comment by kilrau** [comment]
Local `placeorder`
* If an internal match was found (maker order is `ownOrder`), send order invalidation for the maker order (#266)
* If a non-internal match was found (maker order is `peerOrder`), swap is instantiated
    * If swap succeed, do nothing
    * If swap failed, try match again
* If an internal match was found (maker order is `peerOrder`), instantiate swap is instantiated by the remote peer * If swap succeed, send order invalidation * If swap failed, do nothing

**Comment by offerm** [comment]
Need to consider what do between initiating the swap and swap confirmation.

What is the status of the order? where do you see it? what to do with an order that can match it?

Keep in mind that swap can take time, It should take 1 second if everything is good but if there are problems it can take more time. For example, if the peer is not executing the second leg of the swap.

**Comment by sangaman** [comment]
I'm thinking we just have to wait, I think we discussed earlier that once we accept a swap we put the swapped amount of the order "on hold" until the resolution of the swap. But I think it's important here to penalize for delays, more than a few seconds should be cause for concern.

**Comment by moshababo** [comment]
@kilrau 

You covered two cases where `peerOrder` is a taker, but we don't have a scenario like that (`peerOrder` is always a maker), so it's more confusing.

From local perspective, we have only 3 scenarios: 
1) internal match 
2) non-internal local match
3) remote match

**Comment by moshababo** [comment]
@offerm between swap and swap resolution the order is in a limbo. We're not doing anything special with it at the moment, but it's outside of the orderbook, so new incoming orders won't be able to match it.

**Comment by moshababo** [comment]
From our previous discussions I think we concluded that it's not effective because the order can still be taken away after we notify/ask the peer.

**Comment by moshababo** [comment]
From our previous discussions I think we concluded that it's not effective in the current construction, because the order can still be taken away after we notify/ask the peer, and before we instantiate the swap.

**Comment by offerm** [comment]
Guys, The protocol as I created for the SWAP POC and as I now implement in XUD is the follow:

- Starts by the taker (finding  a match)
- Taker sends to the Maker DEAL_REQUEST packet which includes LND's PUBKEY of the taker on the accept payment chain 
- Maker respond with DEAL_RESPONSE packet which includes LND's PubKey of the maker on the accept payment chain, and the r_hash created by the maker 
- Taker sends to Maker SWAP_REQUEST packet which includes the deal id
- Maker initiates the swap
- once the swap is done Maker sends to Taker SWAP_RESPONSE which includes the r_preimage

**Comment by sangaman** [comment]
>  if ownOrder is maker and a new incoming peerOrder is taker, should we notify the peer (one-time) on the match and ask to instantiate the swap?

This is that complicated scenario we'd been talking about where neither side wants to be the "taker." I'd say for the time being we discard peerorders that should match immediately. Asking the other side to begin the swap as taker won't work I think, because in the case that both sides are honest and happened to place near-simultaneous orders what will wind up happening is they'll both notify each other asking the other one to start the swap.

Guys, The protocol as I created for the SWAP POC and as I now implement in XUD is the follow:

```
Starts by the taker (finding a match)
Taker sends to the Maker DEAL_REQUEST packet which includes LND's PUBKEY of the taker on the accept payment chain
Maker respond with DEAL_RESPONSE packet which includes LND's PubKey of the maker on the accept payment chain, and the r_hash created by the maker
Taker sends to Maker SWAP_REQUEST packet which includes the deal id
Maker initiates the swap
once the swap is done Maker sends to Taker SWAP_RESPONSE which includes the r_preimage
```

A couple of thoughts:

1. We can advertise LND pub keys on handshake so that it doesn't have to be shared every time there's a swap - I've been thinking about that for a while and can open a PR for it quickly. We're already advertising Raiden addresses. That way DEAL_REQUEST can just include the order_id and quantity, I believe.
2. What's the purpose of the SWAP_RESPONSE packet? Wouldn't the preimage already be known to the taker at that point?

**Comment by offerm** [comment]
1. It can be done. However, we need to consider cases like LND is not available at the time of handshake or change of LND pubkey. Sending it at time of swap makes things simple. If you wish to do that, let do after Berlin. 

2. The Taker's LND knows it but not the Taker's XUD.

**Comment by sangaman** [comment]
We can share pubkeys at the time of the swap for now. I think down the road, if you change your LND pubkey (which I don't think should happen that often) you'd just need to send new handshake data.

**Comment by offerm** [comment]
@sangaman there is also an aspect of privacy in sharing your pubkey with anyone connected to you (which should be all other exchanges, right?)

The maker may send his pubkey to the taker (DEAL_RESPONSE) only after verifying that there is a valid route between him to the taker and between the taker and itself. If this is not the case, the maker may not agree to the deal and should not share his pubkey with the taker.

**Comment by sangaman** [comment]
That's true, but the xud key can be equally pseudonymous/anonymous to the lightning key. It doesn't necessarily link you to a real identity. Also, I believe even with only sharing pub keys in deal packets, it would still be fairly trivial for a snooper to make sure a valid route exists and then ask to make a deal with no intention of following through. 

When we allow for restrictions on who you connect to though, we could have checks before the handshake is even completed to ensure we're not sharing info and entering deals with any random peer that comes knocking.

**Comment by kilrau** [comment]
Final version, please check @sangaman @moshababo @offerm :

* **Maker order is `ownOrder`** (internal match):
    *  if (taker order is `ownOrder`): execute, send order invalidation for the maker order (#266)
    *  if (taker order is `peerOrder`): *do nothing* (for now, future tracking of this in https://github.com/ExchangeUnion/xud/issues/391)

* **Maker order is `peerOrder`** (external match):
    * if (taker order is `ownOrder`): instantiate swap (@offerm , @sangaman )
        * If swap succeeds, do nothing; 
        * If swap failed, try match again and if no immediate match can be found - add to order book (former taker `ownOrder` becomes a maker order)
    * if (taker order is `peerOrder`): *do nothing* (#242)

**Comment by kilrau** [comment]
"Taker sends to Maker SWAP_REQUEST packet which includes the deal id"

I didn't go through details from the weekend yet, but is this still how we do it? Is there a technical restriction why the taker can't create the preimage and hash and just initiate the swap (sending over the hash? This would save us the SWAP_REQUEST packet. Please update us with the flow here @offerm

**Comment by offerm** [comment]
@kilrau no technical reason. It is like this since you said that the "maker" should control the swap and should be the one to start it.

**Comment by kilrau** [comment]
My bad then. I would like to propose to change the protocol for the taker to start the swap (as in my summary above) and some other improvements to save packets. Continuation in https://github.com/ExchangeUnion/xud/issues/407 since this is not related to the summary which is this issue supposed to give.

**Comment by sangaman** [comment]
Sounds good to me about the 4 scenarios. I made comments about changing the swap procedure in the other PR.

**Comment by kilrau** [comment]
Closing this, will be included in a tech paper update soon.
